src/beanmachine/paic2/include/PaicAST.h

```diff
@@ -0,0 +1,227 @@
+        static bool classof(const Type*c) { return c->getKind() == Primitive; }
+    private:
+        PrimitiveCode _code;
+    };
+
+    class WorldType : public Type {
+    public:
+        WorldType(PrimitiveCode element_type, int length):Type(World), _nodeType(element_type){
+            if(length < 0){
+                _length = 0;
+            } else {
+                _length = length;
+            }
+        }
+        ~WorldType() = default;
+        PrimitiveCode nodeType()const{return _nodeType;}
+        unsigned int length()const{return _length;}
+        static bool classof(const Type*c) { return c->getKind() == World; }
+    private:
+        unsigned int _length;
+        PrimitiveCode _nodeType;
+    };
+
+    class Node {
+    public:
+        static void bind(pybind11::module &m);
+        Node(Location location, NodeKind kind):_location(std::move(location)), _kind(kind) {}
+        virtual ~Node() = default;
+        const Location &loc() { return _location; }
+        const NodeKind getKind() const {return _kind;}
+    private:
+        Location _location;
+        NodeKind _kind;
+    };
+
+    class DeclareValNode : public Node {
+    public:
+        DeclareValNode(Location loc, llvm::StringRef name, NodeKind kind, std::shared_ptr<Type> type):Node(std::move(loc), kind), _name(name),_type(std::move(type)){}
+        std::string getPyName() { return _name; }
+        llvm::StringRef getName() const { return _name; }
+        std::shared_ptr<Type> &getType() { return _type; }
+    private:
+        std::string _name;
+        std::shared_ptr<Type> _type;
+    };
+
+    class ParamNode : public DeclareValNode {
+    public:
+        ParamNode(Location loc, llvm::StringRef name, std::shared_ptr<Type> type): DeclareValNode(std::move(loc), name, Parameter, std::move(type)){}
+    };
+
+    class Expression : public Node {
+    public:
+        Expression(Location loc,NodeKind kind, std::shared_ptr<Type> type):Node(std::move(loc), kind),_type(std::move(type)){}
+        std::shared_ptr<Type> getType() { return _type; }
+    private:
+        std::shared_ptr<Type> _type;
+    };
+
+    class VarNode : public DeclareValNode {
+    public:
+        VarNode(Location loc, llvm::StringRef name, std::shared_ptr<Type> type,
+                std::shared_ptr<Expression> initVal): DeclareValNode(std::move(loc), name, Variable, std::move(type)),_rhs(std::move(initVal)){}
+        std::shared_ptr<Expression> getInitVal() { return _rhs; }
+        static bool classof(const Node*c) { return c->getKind() == paic2::NodeKind::Variable; }
+    private:
+        std::shared_ptr<Expression> _rhs;
+    };
+
+    class GetValNode : public Expression {
+    public:
+        GetValNode(Location loc, llvm::StringRef name, std::shared_ptr<Type> type)
+                : Expression(std::move(loc), NodeKind::GetVal, std::move(type)), _name(name) {}
+
+        std::string getPyName() { return _name; }
+        llvm::StringRef getName() const { return _name; }
```

Comment:
FYI the reason there are two is because the method exposed to Python must return a string (since std::string is mapped to Python str) but the receiver in the builder needs a StringRef. A StringRef is implicitly created from a string so the getName method can safely be deleted